src/server/transaction/server/driver.go

```diff
@@ -225,60 +221,81 @@ func (d *driver) appendTransaction(
-				info.Responses = newResponses
-			}
+		info.Requests = append(info.Requests, items...)
+		return d.runTransaction(txnCtx, info)
+	})
+}
 
-			// This may error with a postgres-level transaction conflict, in which
-			// case we want to save the generated responses and reattempt.  The saved
-			// responses will contain any references to objects generated outside of
-			// the transaction and need to be reused on the next attempt.
-			info, err = d.runTransaction(txnCtx, info)
-			newResponses = info.Responses
+// updateTransaction accepts a function that uses and (possibly) updates a transaction and runs it
+// in either a Read- or WriteContext, taking care to save any new transaction info, except in the case
+// of an unretriable error
+func (d *driver) updateTransaction(
+	ctx context.Context,
+	writeTxn bool,
+	txn *transaction.Transaction,
+	f func(txnCtx *txncontext.TransactionContext, info *transaction.TransactionInfo) (*transaction.TransactionInfo, error),
+) (*transaction.TransactionInfo, error) {
+	// Run this thing in a loop in case we get a conflict, time out after some tries
+	newInfo := new(transaction.TransactionInfo)
+	attempt := func(txnCtx *txncontext.TransactionContext) error {
+		var err error
+		if err = d.transactions.ReadWrite(txnCtx.SqlTx).Get(txn.ID, newInfo); err != nil {
 			return err
-		}); err != nil {
-			return nil, err
 		}
+		if newInfo, err = f(txnCtx, newInfo); err != nil && col.IsErrTransactionConflict(err) {
+			// if we saw ErrTransactionConflict, assume something had to change the transaction response
+			// (such as a pipeline seeing an update outside the transaction), so return this error to
+			// escape the SQL transaction and run the storing code
+			return &transactionModifiedError{}
+		} else {
+			return err
+		}
+	}
 
-		info := &transaction.TransactionInfo{}
-		if err := col.NewSQLTx(ctx, d.db, func(sqlTx *sqlx.Tx) error {
-			// Update the existing transaction with the new requests/responses
-			return d.transactions.ReadWrite(sqlTx).Update(txn.ID, info, func() error {
-				if len(info.Requests) != numRequests || len(info.Responses) != numResponses {
-					// Someone else modified the transaction while we did the dry run
-					return &transactionConflictError{}
-				}
+	for i := 0; i < 10; i++ {
+		var err error
+		if writeTxn {
+			err = d.txnEnv.WithWriteContext(ctx, attempt)
+		} else {
+			err = d.txnEnv.WithReadContext(ctx, attempt)
+		}
+		if err != nil && errors.Is(err, errutil.ErrBreak) {
+			return newInfo, nil // no need to update
+		}
 
-				info.Requests = append(info.Requests, items...)
-				info.Responses = newResponses
-				return nil
-			})
-		}); err == nil {
-			return info, nil
-		} else if !errors.As(err, &transactionConflictError{}) {
+		// a transactionModifiedError from the update function should be understood
+		// as the active transaction needing to change in a way that should be retried after
+		if err == nil || errors.Is(err, &transactionModifiedError{}) {
```

Comment:
why go through update logic when `err == nil `? isn't that condition also implying success

runTransaction called [here](https://github.com/pachyderm/pachyderm/pull/6245/files#diff-4f64d228babb3598928440a7b5e19713aadadc53c352918ee0c12293640ead7eR229) looks like it would return err == nil in a success case